yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -657,6 +657,9 @@ private[spark] class Client(
           case e: ApplicationNotFoundException =>
             logError(s"Application $appId not found.")
             return (YarnApplicationState.KILLED, FinalApplicationStatus.KILLED)
+          case ie: IOException =>
+            logError(s"Application $appId occurred an unexpected IOException.")
+            return (YarnApplicationState.KILLED, FinalApplicationStatus.KILLED)
```

Comment:
@vanzin The process hang due to the endless `renewLease` logic in `DFSClient`, so I think an IOException is just fine for this problem. Other exception may cause the 'monitor' thread down but can not hang the process.
And it seems that if the exception handler is under `run` method, it can't go through the `sc.stop()` logic. If that, the driver process had to wait the network to be reachable then exit.